src/price_data.py

```diff
@@ -104,6 +106,60 @@ def _get_price_binance(self, base_asset: str, utc_time: datetime.datetime, quote
         average_price = total_cost / total_quantity
         return average_price
 
+    @misc.delayed
+    def _get_price_kraken(self, base_asset: str, utc_time: datetime.datetime, quote_asset: str, swapped_symbols: bool = False) -> float:
+        """Retrieve price from Kraken official REST API.
+
+        We select the data point closest to the desired timestamp (utc_time), but not newer than this timestamp.
+        For this we fetch one chunk of the trade history, starting one hour before this timestamp.
+        We then walk through the history until the closest timestamp match is found.
+        Otherwise, we start another 10 minutes earlier and try again, etc. …
+        (Exiting with a warning and zero price after hitting the arbitrarily chosen offset limit of 120 minutes.)
+
+        Documentation: https://www.kraken.com/features/api
+        """
+        target_time = int(utc_time.timestamp() * 1000)
+        root_url = "https://api.kraken.com/0/public/Trades"
+        pair = base_asset + quote_asset
+        pair = kraken_pair_map.get(pair, pair)
+
+        minutes_offset = 0
+        while minutes_offset < 120:
+            minutes_offset += 10
+
+            start_time, end_time = misc.get_offset_timestamps(utc_time, datetime.timedelta(minutes=minutes_offset))
+            since = start_time * 1000000
+            url = f"{root_url}?{pair=:}&{since=:}"
+
+            log.debug(f"Querying trades for {pair} at {utc_time} (offset={minutes_offset}m): Calling %s", url)
+            response = requests.get(url)
+            response.raise_for_status()
+            data = json.loads(response.text)
+
+            if data["error"]:
+                log.warning(f"Querying trades for {pair} at {utc_time} (offset={minutes_offset}m): "
+                            f"Could not retrieve trades: {data['error']}")
+                return 0
+
+            # Find closest timestamp match
+            data = data["result"][pair]
+            closest_match_index = None
+            for i, d in enumerate(data):
+                trade_time = int(float(d[2]) * 1000)
+                time_diff = target_time - trade_time
+                if time_diff >= 0:
+                    closest_match_index = i
+                else:
+                    break
```

Comment:
I believe that the timestamps are sorted? Lets use [bisect](https://docs.python.org/3/library/bisect.html) here.
`closest_match_index = bisect.bisect_left(data, utc_time) - 1` should do the job.

Reply:
Oh boy! Another new function for me. I can certainly look into that, but it'll take some time. I don't really see how that line would work, especially in the context of the for-loop.

Reply:
You do not need a forloop for that. bisect is nothing else than a binary search which should be faster on sorted lists than checking each value.

By the way. We have to confirm, that at least the last value (the latest trade) is after our trade. Otherwhise we would get the wrong price.

Reply:
Thanks for the clarification, I'll look into that tomorrow. Although speed shouldn't really be a big concern here, I'm always happy to learn about other ways to do things!

Yes, I guess you're right that there needs to be some extra logic handling the case of an "extremely dense" trade history where -10 minutes is already too far back.